scarb/src/sources/git/client.rs

```diff
@@ -330,6 +331,51 @@ fn collect_refspecs(reference: &GitReference) -> (Vec<String>, bool) {
     }
 }
 
+/// A wrapper over [`scarb::core::Package`] that provides functionality used to gather VCS info.
+pub struct PackageRepository<'a> {
+    pkg: &'a Package,
+    repo: gix::Repository,
+}
+
+impl<'a> PackageRepository<'a> {
+    pub fn open(pkg: &'a Package) -> Result<Self> {
+        let repo = gix::discover(pkg.root().to_path_buf())?;
+        Ok(Self { repo, pkg })
+    }
+
+    fn work_dir(&self) -> Result<&Path> {
+        self.repo
+            .work_dir()
+            .context("cannot get repository working directory")
+    }
+
+    pub fn is_clean(&self) -> Result<bool> {
+        // `git status -s` output is empty only if there are no changes at all, but always returns success.
+        // `git diff-index --quiet HEAD` returns status code correctly, but doesn't take untracked files into account.
+        let output = git_command()
+            .current_dir(self.work_dir()?)
+            .arg("status")
+            .arg("-s")
+            .output()?;
+
+        Ok(output.stdout.is_empty())
+    }
+
+    pub fn head_rev_hash(&self) -> Result<String> {
+        Ok(self.repo.rev_parse_single("HEAD")?.to_string())
+    }
+
+    /// Calculate relative path from the repository root to the package root.
+    pub fn path_in_vcs(&self) -> Result<String> {
```

Comment:
```suggestion
    pub fn path_in_vcs(&self) -> Result<Utf8PathBuf> {
```

this would make more sense imo